### PR TITLE
Change computation price total of product custom on order detail confirmation

### LINKED ISSUE #10912: Wrong total and total_wt on order-confirmation with customization
**Describe the bug**
in the confirmation page, the total and the total_wt is wrong when we add a customization in the cart
image

**Additionnal information**
PrestaShop version: 1.7.4.2
PHP version: 7.1

In classes\Product.php
image
$product_update['total_wt'] = 13,34  * (1 - 1); -> 0
$product_update['total'] = 13,9  * (1 - 1); -> 0

In templates\checkout\_partials\order-confirmation-table.tpl
image

### PR DESCRIPTION
**Pull request by feyromain**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Bad price total display for custom products on detail order
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #10912 
| How to test?  | Select a customizable product. Add this product to your cart. Confirm your cart. The total price for the customize product is not correct.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/15171)

### CODE REVIEW

**Comment by PierreRambaud** [comment]
Hi and thanks for your contribution.
But I'm not able to reproduce your error.
Can you please create an issue?

**Comment by feyromain** [comment]
Hi,
No problem. I just created an issue #15175 .

File: classes/Product.php
Comment (PierreRambaud) [inline classes/Product.php]:
```suggestion
                    $product_update['total_wt'] = $price_wt * $customization_quantity;
```

File: classes/Product.php
Comment (PierreRambaud) [inline classes/Product.php]:
```suggestion
                    $product_update['total'] = $price * $customization_quantity;
```

**Review by PierreRambaud — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
Thanks I manage to reproduce now :D Thanks again for your contribution.

**Review by lacrm — DISMISSED** [review]

**Comment by sarahdib** [comment]
Hello @feyromain 

Thank you for your contribution.

The pr fix the problem when there is no discount apply to the catalog.
When I create a catalog price rule the price display is not the right one
image
You can see here I have a product with a price to 288,00€ TTC I apply a catalog price rule of 10,00€TTC the price display is the HT amount of the product.

**Comment by feyromain** [comment]
Hello @sarahdib,

Indeed catalog prices were not taken into account.

I change the price used for calculating `total_wt`, `total_customization_wt`, `total` and `total_customization`.

Do you see another problem?

**Comment by sarahdib** [comment]
Hello @feyromain 

I test several combinaison to be sure and seems OK now. 

I'm just waiting for dev review to make last test.

Thank you for your contribution :)

File: classes/Product.php
Comment (PierreRambaud) [inline classes/Product.php:5171]: I think the isset was added for something.
Could you keep the `$price` and `$price_wt` variable instead of using `unit_price_tax_incl` key?

**Comment by marionf** [comment]
Hello @feyromain 
Do you feel to finish this PR ?

**Comment by clotairer** [comment]
I'm interested by the fix of this PR. Can I help you in something ?

**Comment by marionf** [comment]
Hello @clotaire202 
We don't have any news of @feyromain, feel free if you want to finish this PR and handle the returns of @PierreRambaud :)

File: classes/Product.php
Comment (clotairer) [reply classes/Product.php:5171]: In the current release 1.7, product_update come from 3 ways (backtrace):

Products::addCustomizationPrice
Products::addProductCustomizationPrice
Order::getProducts
Order::getProductsDetail > SELECT * FROM `' . _DB_PREFIX_ . 'order_detail` od ...

Products::addCustomizationPrice
Products::addProductCustomizationPrice
OrderInvoice::getProducts
OrderInvoice::getProductsDetail > SELECT * FROM `' . _DB_PREFIX_ . 'order_detail` od ...

Products::addCustomizationPrice
Products::addProductCustomizationPrice
HTMLTemplateOrderSlip::__construct
OrderSlip::getOrdersSlipProducts
Order::getProductsDetail > SELECT * FROM `' . _DB_PREFIX_ . 'order_detail` od ...

So price come from order_details and is already defined.

File: classes/Product.php
Comment (clotairer) [inline classes/Product.php:5194]: $product_update['total_customization'] is not used anywhere in the PrestaShop. I suggest to remove this variable.
in fact $product_update['total'] =  $product_update['total_customization'] ...

File: classes/Product.php
Comment (clotairer) [inline classes/Product.php:5192]: $product_update['total_customization_wt'] is defined only in one template
./admindev/themes/default/template/controllers/carts/helpers/view/view.tpl:109
$product_update['total_customization_wt'] = $product_update['total_wt']
So I suggest to to replace in view.tpl by total_wt and remove $product_update['total_customization_wt']. But since 1.7 this part of template is not used because the cart BO page use PrestaShopBundle...

**Review by clotairer — COMMENTED** [review]
Thank you @marionf

**Comment by clotairer** [comment]
@PierreRambaud 

This part of the code is an heritage of 1.6 and previous release. I've not found cogent usage of total_customization(_wt) : not use in front. Since you refactor the price calculator and ProductPresenter, I this this variable is deprecated.

To be sure, if checked in 1.6. total_customization(_wt) is used in default-bootstrap theme: 
./shopping-cart-product-line.tpl > {if !$priceDisplay}{displayPrice price=$product.total_customization_wt}{else}{displayPrice price=$product.total_customization}{/if}
./js/cart-summary.js > product_customization_total = product_list[i].total_customization_wt;
./order-detail.tpl > {convertPriceWithCurrency price=$product.total_customization_wt currency=$currency}

total_customization_wt is used instead of total in case of customized products. Probably the total(_wt) already is 0. In fact, ($product_quantity - $customization_quantity) always equal 0 because, you cannot add in a cart a product with and without customization in the same product_detail.

What do you think about my suggestion ?
This PR has more than one year. Perhaps I need to create a new one to avoid conflict ?

**Comment by mabuch** [comment]
@clotairelims 
I have this issue with customizations as well. For some time I fixed it in src/Adapter/Presenter/Order/OrderLazyArray.php (see also issue #12264 where this is mentioned as well). Today I found this issue and I have created an override of Product.php for the addCustomizationPrice() function as suggested by @feyromain (as only fixing the output in OrderLazyArray.php proved insufficient for another problem I ran into today, therefore I started looking for a better solution which tackles the problem at the source).

After reading your comments I also decided to try not setting 'total_customization_wt' and 'total_customization' as suggested (deleting lines 5192 and 5194). A first test looks very promising, the issue is fixed and so far I have not run into any problems due to emitting 'total_customization_wt' and 'total_customization'. I will run additional tests in the upcoming days, but I think your proposition should be suitable.

**Comment by mabuch** [comment]
Update: total_customization_wt cannot yet be removed (line 5192) in classes/Product.php.
The reason is, that PrestaShop/src/Adapter/Cart/QueryHandler/GetCartForViewingHandler.php imlicitly relys on it being set when calling formatPrice() in  prepareProductForView() [line 238]. Without it being present the price formatting fails (as there is no price) and in the Admin Interface it is no longer possible to open Carts that contain customisations (not Orders, they still work fine, but Carts are broken).

**Review by matks — CHANGES_REQUESTED** [review]
Hello everyone. Sorry for the delay, it seems this PR got lost in the abyss :( .

I read the two issues related to this PR. I also read the discussions and the information provided by clotairelims and @PierreRambaud. It seems this part is fragile and it's very hard to know whether a variable value can be changed without side-effects on other parts of the code.

Without an automated test (unit test or integration test) that
- fails before this change
- then we add the change
- then the test is green

I cannot validate this PR. Sorry 😞  .

**Comment by clotairer** [comment]
Hi @matks
I'm back... How can I help you ?

**Comment by matks** [comment]
> Hi @matks I'm back... How can I help you ?

I thought you were gone to the Java world 😄

**Comment by metainnova** [comment]
Hi to everybody,

I have took a look to a lot of related tickets speaking about that issue, but for I have seen, the issue has not been solved yet. We have installed a new shop, and when the product is customized, the total seen in that line of product is always 0,0€. 

Any solution at source code to fix this without secondary effects?

Thanks a lot.

**Comment by matks** [comment]
So, as I mentioned, this PR is hard to merge into the project.

It's not because the changes are bad. It's because the stuff written in this class by developers long gone is undocumented, untested, quite unreadable (it's a class that has 8,000 lines!) and we have no information what the expected behavior was. Also this could be linked to other parts of code, same we have no idea.

In order to be able to approve this PR I need to know how I can be sure this will not introduce a bug. And I don't know how I can be sure of this 😭 .

**Comment by mabuch** [comment]
@matks When I looked through the code I had the impression, that there are multiple versions on how to calculate prices (code duplication). I do not know all the places where prices are calculated, and therefore I cannot tell you how many different versions of price calculations there are.
The order (detail) confirmation has (from what I remember, this might be false) no special/additional calculations when compared to the invoice or the calculation in the backoffice. Therefore it might be possible to look at the price calculations there in order to determine the expected behaviour. Because if I remember correctly, this is only an issue on the order confirmation page, Invoice and Backoffice are correct (but I might be mistaken, please check for yourself). See also #12264, which talks about backward compatibility code. I do not know how much backward compatibility you generally want (and still can) provide, but due to the fact that 1.7 now is several years old and has undergone some bigger changes, my guess is that this legacy code could (and should) be eliminated. The main goal with price calculations within prestashop should be consistency, no matter where the prices are displayed.

Edit: I am aware that a proper cleanup will be more work than trying to quickly fix this issue. But as a quick fix does not seem to be an option and a lot of analysis needs to be done anyways, this could be taken as an opportunity to clean up the code, get rid of legacy calculations and simplify it. I myself lack the knowledge to undertake something on this scale, I have no deeper knowledge of prestashop, I just punctually look at the code when there is an issue and try to fix it. But I can assist you with reviewing and testing (to the best of my capabilities).

**Comment by MatShir** [comment]
@matks what is the state of the PR ? It is fixing a major bug, it's a shame to see it blocked like that. Say what you need to make it move forward, test scenarios forward, better unit test coverage and a second manual qa test, should we ask help from auto qa dev team, should we close this PR ?

**Comment by matks** [comment]
@MatShir As I said:

> And I don't know how I can be sure of this 😭 .

And the same is probably true for other maintainers which is nobody is answering on this PR.

I honestly have no idea what to do 🤷‍♀️  . The stuff written in this class by developers long gone is undocumented, untested, quite unreadable and we have no information what the expected behavior was. It's gibberish.

Maybe ping some other people, sometimes other opinions provide better insights. I'm speechless here.

**Comment by MatShir** [comment]
Then I guess we can close this PR, the content might be useful to some people 🤷 
Thanks, @matks for the investigation

**Comment by clotairer** [comment]
We keep an override on our merchants until when?
If anyone want take the risk to fix a bug concerning the price computation of this "magic ghost" feature. 
Are you killing the custom products feature?

**Comment by MatShir** [comment]
@matks could we introduce the bug fix into a feature flag and try to see the impact afterwards?

**Comment by clotairer** [comment]
In this PR you have thwo things : 
- a refactoring attempt
- a fix between an if ($customization_quantity) < I imagine this is not risky...

Can we fix the bug and postpone the refactoring ?

**Comment by vttn** [comment]
@matks can you please let us know when this bug will be addressed. Regards.

**Comment by matks** [comment]
@vttn I don't know, unfortunately.

@clotairer In order to find an exit to this very very complex situation, @camilleganzin has started the massive work of exploring and understanding the mechanisms behind. I hope thanks to her work, we will be able to handle these contributions nicely. I think she might be happy to have knowledge about this topic from your point of view.

**Pull request closed without merging**
